### PR TITLE
Watermark images now display in the proper aspect ratio in the UI

### PR DESCRIPTION
**Pull request by int-Frank**
[AB#843](https://dev.azure.com/euclideon/57c3bd2a-8f94-4578-b195-2e4fa9d8f295/_workitems/edit/843)

### CODE REVIEW

File: src/vcSettingsUI.cpp [hunk 1]
```diff
@@ -517,7 +517,29 @@ void vcSettingsUI_Show(vcState *pProgramState)
...
+                if (dimension[i] > maxDimension[i])
+                {
+                  float factor = float(dimension[i]) / maxDimension[i];
+                  dimension[i] = int(float(dimension[i]) / factor);
+                  int j = (i + 1) % 2;
+                  dimension[j] = int(float(dimension[j]) / factor);
+
+                  //Is the other dimension too thin now?
+                  if (dimension[j] < minDimension)
+                    dimension[j] = minDimension;
```
Comment (samsinsane) [inline src/vcSettingsUI.cpp]: Doesn't this break the aspect ratio? Though I guess anyone with a `512+:1` or `1:256+` ratio "watermark" isn't really going to have a fun time at all.

Comment (int-Frank) [reply src/vcSettingsUI.cpp]: Yes it does. My intent is to display something, even if the aspect ratio of the image is such that one dimension vanishes after resizing to fit in the UI. Perhaps another solution could be to simply enlarge and clip such images, thus only showing a portion of the watermark, but still retaining the correct aspect ratio?

File: src/vcSettingsUI.cpp [hunk 2]
```diff
@@ -517,7 +517,29 @@ void vcSettingsUI_Show(vcState *pProgramState)
 
             if (pProgramState->settings.convertdefaults.watermark.pTexture != nullptr)
             {
-              ImGui::Image(pProgramState->settings.convertdefaults.watermark.pTexture, ImVec2(256, 256));
+              //Since we're allowing images of any dimensions, we need to make sure it fits in the UI.
+              int dimension[2] = { pProgramState->settings.convertdefaults.watermark.width,
```
[the remark this answers is not in the document]
Comment (int-Frank) [reply src/vcSettingsUI.cpp]: Righto

File: src/vcSettingsUI.cpp [hunk 3]
```diff
@@ -517,7 +517,29 @@ void vcSettingsUI_Show(vcState *pProgramState)
...
+                                   pProgramState->settings.convertdefaults.watermark.height };
+
+              int maxDimension[2] = {512, 256};
```
[the remark this answers is not in the document]
Comment (int-Frank) [reply src/vcSettingsUI.cpp]: Righto

File: src/vcSettingsUI.cpp [hunk 4]
```diff
@@ -517,7 +517,29 @@ void vcSettingsUI_Show(vcState *pProgramState)
...
+              int minDimension = 2;
+
+              for (int i = 0; i < 2; i++)
```
[the remark this answers is not in the document]
Comment (int-Frank) [reply src/vcSettingsUI.cpp]: Good point

File: src/vcSettingsUI.cpp [hunk 5]
```diff
@@ -517,7 +517,29 @@ void vcSettingsUI_Show(vcState *pProgramState)
...
+              {
+                if (dimension[i] > maxDimension[i])
+                {
+                  float factor = float(dimension[i]) / maxDimension[i];
+                  dimension[i] = int(float(dimension[i]) / factor);
+                  int j = (i + 1) % 2;
```
[the remark this answers is not in the document]
Comment (int-Frank) [reply src/vcSettingsUI.cpp]: Good point

File: src/vcSettingsUI.cpp (hunk 1 above)
Comment (int-Frank) [reply src/vcSettingsUI.cpp]: Also note this is only for the preview in settings, not the actual watermark

Comment (samsinsane) [reply src/vcSettingsUI.cpp]: I'm not too fussed on this - I'm not sure it's a realistic situation anyway. This comment probably didn't need to be made, I just wanted confirmation that the aspect ratio was intentionally broken.

File: src/vcSettingsUI.cpp [hunk 6]
```diff
@@ -517,7 +517,26 @@ void vcSettingsUI_Show(vcState *pProgramState)
 
             if (pProgramState->settings.convertdefaults.watermark.pTexture != nullptr)
             {
-              ImGui::Image(pProgramState->settings.convertdefaults.watermark.pTexture, ImVec2(256, 256));
+              //Since we're allowing images of any dimensions, we need to make sure it fits in the UI.
+              udInt2 dimension = {pProgramState->settings.convertdefaults.watermark.width, pProgramState->settings.convertdefaults.watermark.height};
```
Comment (samsinsane) [inline src/vcSettingsUI.cpp]: Technically you create these `udInt2::create` but this is fine.

**Pull request merged**